[Zen2] PersistedState interface implementation

andrershov opened the pull request:
Today `GatewayMetaState` is capable of atomically storing `MetaData` to
disk. We've also moved fields that are needed to be persisted in Zen2
from `ClusterState` to `ClusterState.MetaData.CoordinationMetaData`.

This PR implements `PersistedState` interface. 

1. `version` and `currentTerm` are persisted as a part of `Mainfest`.
2. `GatewayMetaState` now have two descendants `Zen1GatewayMetaState`
and `Zen2GatewayMetaState` (which implements `PersistedState`).
3. `GatewayMetaState` now constructs `previousClusterState` (including
`MetaData`) and `previousManifest` inside the constructor, so that all
`PersistedState` methods are usable as soon as `GatewayMetaState`
instance is constructed. Also `loadMetaData` is renamed to
`getMetaData`, because it just returns
`previousClusterState.metaData()`.
4. Currently when deciding whether to write `IndexMetaData` to disk,
we're comparing current `IndexMetaData` version and received
`IndexMetaData` version. This is not safe in Zen2 if term has changed.
So `updateClusterState` now accepts `writeWithoutComparingVersion`
method parameter. When it's set to true, we always write
`IndexMetaData` to disk. 
5. Things that are not covered by `GatewayMetaStateTests` are covered
by `Zen2GatewayMetaStateTests`.

This PR does not inject `Zen2GatewayMetaState` in place of
`Zen1GatewayMetaState` and in place of `InMemoryPersistedState`, this
is a work to be done in a follow-up PR.

elasticmachine commented:
Pinging @elastic/es-distributed

ywelsch commented on server/src/main/java/org/elasticsearch/gateway/Zen2GatewayMetaState.java:
```diff
@@ -0,0 +1,77 @@
+        try {
+            boolean writeWithoutComparingVersions = previousClusterState.term() != clusterState.term();
+            updateClusterState(clusterState, previousClusterState, writeWithoutComparingVersions);
+            previousClusterState = clusterState;
```
why set this again? That's already done by updateClusterState?

andrershov replied:
https://github.com/elastic/elasticsearch/pull/35819/commits/5afb5ee2f0f79f73cb59d73f929446977b4949e3

ywelsch commented on server/src/main/java/org/elasticsearch/node/Node.java:
```diff
@@ -473,6 +471,8 @@ protected Node(
             ).collect(Collectors.toSet());
             final TransportService transportService = newTransportService(settings, transport, threadPool,
                 networkModule.getTransportInterceptor(), localNodeFactory, settingsModule.getClusterSettings(), taskHeaders);
+            final GatewayMetaState gatewayMetaState = new Zen1GatewayMetaState(settings, nodeEnvironment, metaStateService,
```
if we want to flexibly run with Zen1 or Zen2, we need to make this somewhat configurable. This can be handled in a follow-up, however.

ywelsch commented on server/src/test/java/org/elasticsearch/gateway/Zen2GatewayMetaStateTests.java:
```diff
@@ -0,0 +1,235 @@
+            final long currentTerm = randomNonNegativeLong();
+            gateway.setCurrentTerm(currentTerm);
+            gateway = maybeNew(gateway);
+            assertThat(gateway.getCurrentTerm(), equalTo(currentTerm));
```
also check that the cluster state remains untouched?

andrershov replied:
See below

ywelsch commented on server/src/test/java/org/elasticsearch/gateway/Zen2GatewayMetaStateTests.java:
```diff
@@ -0,0 +1,235 @@
+            gateway = maybeNew(gateway);
+            assertThat(gateway.getLastAcceptedState().version(), equalTo(version));
+            assertTrue(MetaData.isGlobalStateEquals(gateway.getLastAcceptedState().metaData(), metaData));
+            assertThat(gateway.getLastAcceptedState().metaData().index(indexName), equalTo(indexMetaData));
```
check that current term is unchanged?

andrershov replied:
See below

ywelsch commented on server/src/test/java/org/elasticsearch/gateway/Zen2GatewayMetaStateTests.java:
```diff
@@ -0,0 +1,235 @@
+        assertThat(gateway.getLastAcceptedState().metaData().index(indexName), equalTo(newIndexMetaData));
+    }
+
+    public void testCurrentTermAndTermAreDifferent() throws IOException {
```
what's the benefit of this test? By asserting what's not changed in the above tests. I think that it also covers this test?

andrershov replied:
Replying to this and two comments above. I just want to make `testSetCurrentTerm` to test only `seCurrentTerm` and `testSetLastAcceptedState` to test only `setLastAcceptedState`. As you may notice `testSetCurrentTerm` and `testSetLastAccepted` tests do not event call `setLastAcceptedState` and `setCurrentTerm`, they focus exclusively on testing their methods. This test focuses on these two methods interaction because it's explicitly required by JavaDoc.

ywelsch commented on server/src/test/java/org/elasticsearch/gateway/Zen2GatewayMetaStateTests.java:
```diff
@@ -0,0 +1,235 @@
+        gateway.markLastAcceptedConfigAsCommitted();
+
+        gateway = maybeNew(gateway);
+        assertThat(gateway.getLastAcceptedState().getLastAcceptedConfiguration(),
```
also assert that nothing else has changed?

andrershov replied:
https://github.com/elastic/elasticsearch/pull/35819/commits/971aef183925c61ee6b147ae5102ad951e3a52b8

andrershov commented on server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -63,45 +63,55 @@
         ensureAtomicMoveSupported(); //TODO move this check to NodeEnvironment, because it's related to all types of metadata
         upgradeMetaData(metaDataIndexUpgradeService, metaDataUpgrader);
-        profileLoadMetaData();
+        initializeClusterState(ClusterName.CLUSTER_NAME_SETTING.get(settings), transportService.getLocalNode());
```
@ywelsch any ideas where can we get localNode in this case?

andrershov commented on server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -255,39 +243,44 @@ void rollback() {
-     *                     which is subclass of {@link IOException} explicitly. See also {@link WriteStateException#isDirty()}.
+     * @param newState new {@link ClusterState}
+     * @param previousState previous {@link ClusterState}
+     * @param writeWithoutComparingVersion if set to true even if previous {@link MetaData} and {@link IndexMetaData} have the same versions
```
https://github.com/elastic/elasticsearch/pull/35819/commits/52fac02d944b545fec0b6744c17b416cdaae173a

andrershov commented on server/src/test/java/org/elasticsearch/gateway/Zen2GatewayMetaStateTests.java:
```diff
@@ -0,0 +1,235 @@
+import static org.hamcrest.Matchers.not;
+
+public class Zen2GatewayMetaStateTests extends ESTestCase {
+    private class Gateway extends Zen2GatewayMetaState {
```
Gateway -> Zen2GatewayMetaStateUT https://github.com/elastic/elasticsearch/pull/35819/commits/6de8c9a2fe857ea6d132d0aa92e0868a64b34f7c

andrershov commented on server/src/main/java/org/elasticsearch/cluster/metadata/Manifest.java:
```diff
@@ -44,16 +44,28 @@
 public class Manifest implements ToXContentFragment {
-    private static final long MISSING_GLOBAL_GENERATION = -1;
+    private static final long MISSING_GLOBAL_GENERATION = -1L;
+    private static final long MISSING_CURRENT_TERM = 0L;
```
https://github.com/elastic/elasticsearch/pull/35819/commits/89e14215e1955eccb0d0e77451ff92da795c929c

andrershov commented:
@ywelsch @DaveCTurner, I've made further changes, could you please take a look?

1. I've merged `Zen1GatewayMetaState` and `Zen2GatewayMetaState` into `GatewayMetaState` to make sure we can use GatewayMetaState for both Zen1 and Zen2 discovery. 
2. GatewayService is no longer responsible for registering `GatewayMetaState` as `ClusterStateApplier`, now `ZenDiscovery` is responsible for that. 
3. `ZenDiscovery` accepts `ClusterApplier`, not `ClusterApplierService`. And it turned out to be not easy to make it accept `ClusterApplierService` and make tests working. That's why I've moved `addLowPriorityApplier` to `ClusterApplier` interface, which is not elegant.
4. My first attempt was to change `TestZenDiscovery` (in zen2 mode) to use `GatewayMetaState` instead of `InMemoryPersistedState`, after this change 10 integration tests were failing, because we need to implement proper state recovery, which is a subject of separate PR. That's why I've implemented another option to use `GatewayMetaState` in place of `InMemoryPersistedState` with `InMemoryPersistedState` being default. If`InMemoryPeristedState` is used, we're currently registering `addLowPriorityApplier` with `GatewayMetaState`, to make sure it receives all cluster state updates (otherwise 13 integration tests and one unit tests are failing).
5. There is one test that uses Zen2 with `GatewayMetaState`, see `PersistedStateIT`. It tests that cluster persistent settings work properly, it does not pass with `InMemoryPersistentState`. I was trying to make a test with `IndexMetaData`, but it requires proper recovery implementation.

ywelsch commented on server/src/test/java/org/elasticsearch/gateway/Zen2GatewayMetaStateTests.java:
```diff
@@ -0,0 +1,246 @@
+    private class GatewayMetaStateUT extends GatewayMetaState {
+        GatewayMetaStateUT(Settings settings, NodeEnvironment nodeEnvironment) throws IOException {
+            super(settings, nodeEnvironment, new MetaStateService(nodeEnvironment, xContentRegistry()),
+                    Mockito.mock(MetaDataIndexUpgradeService.class), Mockito.mock(MetaDataUpgrader.class));
```
just null?

andrershov replied:
I generally prefer to avoid nulls.

ywelsch commented on test/framework/src/main/java/org/elasticsearch/test/discovery/TestZenDiscovery.java:
```diff
@@ -76,29 +80,40 @@ public TestPlugin(Settings settings) {
+                            return gatewayMetaState;
+                        };
+                    } else {
+                        //if InMemoryPersisted is used, we let GatewayMetaState receive all events
```
... because some tests rely on it due to dangling indices functionality.

andrershov replied:
https://github.com/elastic/elasticsearch/pull/35819/commits/d91ad3cab253cf519ec906445e7976c1be26d8d3

ywelsch commented on server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -90,18 +90,29 @@ public GatewayMetaState(Settings settings, NodeEnvironment nodeEnv, MetaStateSer
     }
 
-    private void upgradeMetaData(MetaDataIndexUpgradeService metaDataIndexUpgradeService, MetaDataUpgrader metaDataUpgrader)
+    public void setLocalNode(DiscoveryNode localNode) {
```
can you add an assertion that this is only called once?

andrershov replied:
https://github.com/elastic/elasticsearch/pull/35819/commits/35d38872847458bd2170c4003babc5063c78c2e1

ywelsch commented on server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -164,22 +176,55 @@ public void applyClusterState(ClusterChangedEvent event) {
+
+    @Override
+    public ClusterState getLastAcceptedState() {
+        assert previousClusterState.nodes().getSize() > 0 : "Call setLocalNode before calling this method";
```
why not check that `getLocalNode() != null`? It's a stronger check than what's currently being checked here.

andrershov replied:
https://github.com/elastic/elasticsearch/pull/35819/commits/24c43fa135b603e1e2c5dca70182a80a3e05d35d

ywelsch commented on server/src/test/java/org/elasticsearch/gateway/Zen2GatewayMetaStateTests.java:
```diff
@@ -0,0 +1,246 @@
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.not;
+
+public class Zen2GatewayMetaStateTests extends ESTestCase {
```
remove the Zen2 prefix of this test? It's testing the `GatewayMetaState` class

andrershov replied:
There is already `GatewayMetaStateTests` class in this package, it tests common functionality for both `zen1` and `zen2` and it has a different structure from this class (it avoids creating `GatewayMetaState` instance for each test). I've renamed this class to `GatewayMetaStatePersistedStateTests` to make it clear, that we're checking how `GatewayMetaState` implements `PersistedState` interface methods.

andrershov replied:
https://github.com/elastic/elasticsearch/pull/35819/commits/7dbb9af3cb9416ac280e4d59f8cd5a7947778a31

andrershov commented on server/src/main/java/org/elasticsearch/cluster/service/ClusterApplier.java:
```diff
@@ -57,4 +58,6 @@ default void onSuccess(String source) {
         void onFailure(String source, Exception e);
     }
+
+    void addLowPriorityApplier(ClusterStateApplier applier);
```
https://github.com/elastic/elasticsearch/pull/35819/commits/2181474a3634b521452ad6c457ff6c7acdda5fce

andrershov commented on server/src/main/java/org/elasticsearch/discovery/DiscoveryModule.java:
```diff
@@ -118,15 +123,22 @@ public DiscoveryModule(Settings settings, ThreadPool threadPool, TransportServic
             () -> new ZenDiscovery(settings, threadPool, transportService, namedWriteableRegistry, masterService, clusterApplier,
-                clusterSettings, hostsProvider, allocationService, Collections.unmodifiableCollection(joinValidators)));
+                clusterSettings, hostsProvider, allocationService, Collections.unmodifiableCollection(joinValidators), gatewayMetaState));
+        discoveryTypes.put("zen2", () -> new Coordinator(NODE_NAME_SETTING.get(settings), settings, clusterSettings, transportService,
```
https://github.com/elastic/elasticsearch/pull/35819/commits/95a788fe7a1c30fa84144599949188267ebc20d2

andrershov commented on server/src/main/java/org/elasticsearch/node/Node.java:
```diff
@@ -473,6 +470,8 @@ protected Node(
             ).collect(Collectors.toSet());
             final TransportService transportService = newTransportService(settings, transport, threadPool,
                 networkModule.getTransportInterceptor(), localNodeFactory, settingsModule.getClusterSettings(), taskHeaders);
+            final GatewayMetaState gatewayMetaState = new GatewayMetaState(settings, nodeEnvironment, metaStateService,
```
https://github.com/elastic/elasticsearch/pull/35819/commits/5bdff70ef0fe54a1f1fea2a5e225df998b930bc5

andrershov commented on server/src/test/java/org/elasticsearch/cluster/coordination/PersistedStateIT.java:
```diff
@@ -0,0 +1,57 @@
+        .build();
+
+
+    public void testPersistentSettingsOnFullRestart() throws Exception {
```
Please see my point 5 above. I was trying to test IndexMetaData as well, but this requires proper recovery implementation. Also, this test passes with InMemoryPersistedState, because of dandling indices. I'll add this test once recovery is properly implemented.

andrershov commented on server/src/test/java/org/elasticsearch/discovery/zen/ZenDiscoveryUnitTests.java:
```diff
@@ -357,12 +359,17 @@ public void setInitialState(ClusterState initialState) {
             masterService, clusterApplier, clusterSettings, hostsResolver -> Collections.emptyList(),
             ESAllocationTestCase.createAllocationService(),
-            Collections.emptyList());
+            Collections.emptyList(), mock(GatewayMetaState.class));
```
I generally prefer to avoid nulls, here it's clear that we're passing an instance of `GatewayMetaState` and methods of this instance do nothing.

ywelsch replied:
I find it simpler to use `null`, as it makes it clear that none of the methods on this object are invoked, but that's a personal preference.

andrershov commented:
@ywelsch I'm done with the changes, ready for the next pass.
